### ISSUE TITLE
Bug: IPv6 doesn't always include last 4 parts.

### ISSUE THREAD

**Issue by charlesomer** [qdm12/ddns-updater#357]
**TLDR**: IPv6 address sometimes gets set without the last 4 parts.

1. Is this urgent: Yes
2. DNS provider(s) you use: Digital Ocean
3. Program version: Running version latest built on 2022-03-15T14:38:50Z (commit 584597d)

    

    `Running version latest built on 2020-03-13T01:30:06Z (commit d0f678c)`

4. What are you using to run the container: docker-compose
5. Extra information (optional): I have tried with both /64 and /128. I can't see a way to exec into the container to test the curls but they work fine from the host. Apologies if I've missed something obvious. It's odd to me that it will sometimes work correctly.

Logs:
Due to restarting the container I don't have the logs at the moment.

Configuration file (**remove your credentials!**):

```json
  ddns-updater:
    image: qmcgaw/ddns-updater
    container_name: ddns-updater
    network_mode: bridge
    ports:
      - 8000:8000/tcp
    environment:
      - CONFIG={"settings":[{"provider":"digitalocean","domain":"[removed]","host":"[removed]","token":"[removed]","ip_version":"ipv6"}]}
      - BACKUP_PERIOD=72h # 0 to disable
      - TZ=Europe/London
    restart: always
    logging:
      options:
        max-size: "200k"
        max-file: "10"
```

Host OS:
Raspberry Pi OS 32 bit

**Comment by qdm12** [comment]
> No change, except that setting `PUBLICIP_FETCHERS=dns` now results in the same output as setting `http`.

Fixed in 81b336fe0719ff3224c70692f4ae37e086cde927 and 6513e5869adf62faa74a68b2caf99bd73fd73066 my bad that was a regression introduced with the *"Environment variables parsing was changed on 12 June, please report any issue you might have"*

What do you get with:

```sh
docker run --rm alpine:3.18 wget -qO- https://api6.ipify.org
docker run --rm alpine:3.18 wget -qO- http://ip1.dynupdate6.no-ip.com
```

These two urls are for IPv6 only by HTTP. Maybe it's just their response which is wrong 🤔 
And also out of curiosity (this one is for ipv4 or ipv6), what do you get from:

```sh
docker run --rm alpine:3.18 wget -qO- https://domains.google.com/checkip
```

? Thanks!! 👍

**Comment by qdm12** [comment]
Solved it for http: 21be712d845228f2d62790d0429474058548e1a8

TLDR: used a bad regex from a ihateregex.com which really makes me hate regex and this website.

@Ryuukyu313 can you create an issue with the logs running the latest ddns image to see why it complains about a mismatching ip address (`When using dns instead, this happens:`) thank you!

**Comment by qdm12** [comment]
I will finish fixing it tomorrow (it's 11pm here 💤), but yeah the ipv6 regex is bad, and I can't find a decent one (and don't want to do a phd on regex either), so I'm writing my own ipv6 extraction function which will be more reliable and tested.

**Comment by qdm12** [comment]
Fixed in 954dffd3a78751061c01e0f9135086a306f57dc2

**Comment by Ryuukyu313** [comment]
@qdm12 Can confirm https://github.com/qdm12/ddns-updater/commit/954dffd3a78751061c01e0f9135086a306f57dc2 fixes the issue, the correct IP address is now updated.

The mismatch error when using the dns fetcher is also no longer present, it works just fine now.
Looking at the old logs again, I'd suspect that the same faulty regex was used on the response from googledomains,
which then caused the mismatch.

And thanks a lot for your work! Out of all ddns clients I've tried, yours is now the only one I could get to work with my IPv6-only setup.
